### PR TITLE
Rebase virtual package and state handling

### PR DESCRIPTION
**Pull request by Mikkelhost**

### CODE REVIEW

File: internal/state/models.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	VboxOpts      []exercise.ExerciseInstanceConfig
+	Tag           string
+	Net           *virtual.Network
+	DnsAddr       string
+	DnsRecords    []exercise.RecordConfig
+	Ips           []int
+	Containers    []*virtual.Container
+	Vms           []*virtual.Vm
+	// Vms           []Vm
+	// Containers    []Container
```
Comment (Lanestolen) [inline internal/state/models.go]:
```suggestion

```

File: internal/agent/agent.go [hunk 2]
```diff
@@ -153,8 +147,13 @@ func New(conf *Config) (*Agent, error) {
...
 		log.Info().Msg("No redis container detected, creating a new one")
-		container = docker.NewContainer(docker.ContainerConfig{
-			Image:     "redis:7.0.4",
+		container = virtual.NewContainer(virtual.ContainerConfig{
+			Image: "redislabs/rejson:2.4.1",
+			// Cmd: []string{
+			// 	"/data/redis.conf",
+			// 	"--loadmodule",
+			// 	"/usr/lib/redis/modules/rejson.so",
+			// },
```
Comment (Lanestolen) [inline internal/agent/agent.go]:
```suggestion
```
Don't comment out commented lines, this is one way to expose secrets by accident :)

File: internal/agent/agent.go [hunk 3]
```diff
@@ -30,19 +28,15 @@ var configPath string
 type Agent struct {
 	initialized bool
 	config      *Config
-	redis       cache.RedisCache
-	State       *State
+	redis       state.RedisCache
+	State       *state.State
 	auth        Authenticator
-	vlib        vbox.Library
+	vlib        *virtual.VboxLibrary
```
Comment (Lanestolen) [inline internal/agent/agent.go:32]: Not sure why you made this change in the virtual

File: internal/agent/environment.go [hunk 4]
```diff
@@ -125,7 +127,7 @@ func (a *Agent) CloseEnvironment(ctx context.Context, req *proto.CloseEnvRequest
 	vpnIP := strings.ReplaceAll(envConf.VPNAddress, ".240.1/22", "")
 	vpnIPPool.ReleaseIP(vpnIP)
 
-	if err := vbox.RemoveEventFolder(string(envConf.Tag)); err != nil {
+	if err := virtual.RemoveEventFolder(string(envConf.Tag)); err != nil {
 		//do nothing
 	}
```
Comment (Lanestolen) [inline internal/agent/environment.go]:
```suggestion
virtual.RemoveEventFolder(string(envConf.Tag))
```
Or at least log the error :)

File: internal/environment/guac.go [hunk 5]
```diff
@@ -65,27 +63,27 @@ func NewGuac(ctx context.Context, eventTag string) (Guacamole, error) {
 Creates the necessary containers for guacamole and configures the instance with a new admin password
 */
 func (guac *Guacamole) create(ctx context.Context, eventTag string) error {
-	_ = vbox.CreateEventFolder(eventTag)
+	_ = virtual.CreateEventFolder(eventTag)
```
Comment (Lanestolen) [inline internal/environment/guac.go]:
```suggestion
	virtual.CreateEventFolder(eventTag)
```

File: internal/environment/iptables.go [hunk 6]
```diff
@@ -10,19 +10,19 @@ import (
 )
 
 type IPTables struct {
```
Comment (Lanestolen) [inline internal/environment/iptables.go:12]: Why export everything like this?

File: internal/environment/lab/lab.go [hunk 7]
```diff
@@ -136,14 +132,15 @@ func (l *Lab) Start(ctx context.Context) error {
 func (l *Lab) Close() error {
 	var wg sync.WaitGroup
 	for _, lab := range l.Frontends {
+		log.Debug().Msgf("lab: %v", lab)
```
Comment (Lanestolen) [inline internal/environment/lab/lab.go]: Hmm what is this doing here, add more context to the statement and use fields :)

File: internal/environment/lab/network/dhcp/dhcp.go [hunk 8]
```diff
@@ -69,40 +69,40 @@ func New(format func(n int) string) (*Server, error) {
...
 	return &Server{
-		cont:     cont,
-		confFile: confFile,
-		dns:      dns,
-		subnet:   subnet,
+		Cont:     cont,
+		ConfFile: confFile,
+		Dns:      dns,
+		Subnet:   subnet,
 	}, nil
```
Comment (Lanestolen) [inline internal/environment/lab/network/dhcp/dhcp.go:76]: Why export everything?

File: internal/environment/iptables.go (hunk 6 above)
Comment (Mikkelhost) [reply internal/environment/iptables.go:12]: It is needed when saving the state

File: internal/environment/lab/lab.go (hunk 7 above)
Comment (Mikkelhost) [reply internal/environment/lab/lab.go]: Had some errors that i didn't know where came from, so just used it for analyzing the problem. Removed in next commit

File: internal/environment/lab/network/dhcp/dhcp.go (hunk 8 above)
Comment (Mikkelhost) [reply internal/environment/lab/network/dhcp/dhcp.go:76]: Same reason for iptables, needed when saving the state

File: internal/agent/agent.go (hunk 3 above)
Comment (Mikkelhost) [reply internal/agent/agent.go:32]: Removed the interfaces which made no sense to have

Comment (Mikkelhost) [reply internal/agent/agent.go:32]: Also gathered docker and vbox under the virtual package.
May make sense to split them into their own packages again for readability

**Pull request merged**